### PR TITLE
CDAP-6595 Authorization for Metadata

### PR DESCRIPTION
**Pull request by rohitsinha54**
Issue: https://issues.cask.co/browse/CDAP-6595

- Added authorization for Metadata (Tags and Properties)

Build: http://builds.cask.co/browse/CDAP-RUT330-1
ITM: http://builds.cask.co/browse/CDAP-ITM6-7

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java [hunk 1]
```diff
@@ -174,6 +200,27 @@ public void removeTags(NamespacedEntityId namespacedEntityId, String... tags) th
...
 
   /**
+   * Ensures that the principal making the request has privilege to perform any one {@link Action}
+   * on the given namespacedEntityId
+   *
+   * @param namespacedEntityId the {@link NamespacedEntityId} on which privilege needs to be ensured
+   * @throws UnauthorizedException if the principal making the request does not have privilege to
+   * perform even one {@link Action} on the given namespacedEntityId
+   */
+  private void ensureAccess(NamespacedEntityId namespacedEntityId) throws UnauthorizedException {
```
Comment (Nishith) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java:210]: s/ensureAccess/ensureAnyAccess/ ?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java [hunk 2]
```diff
@@ -72,15 +83,103 @@ public static void setup() throws Exception {
...
+      Assert.assertTrue(e instanceof UnauthorizedException);
+    }
+
+    // Grant permission to Bob on entity
+    SecurityRequestContext.setUserId(ALICE.getName());
+    authorizer.grant(appId, BOB, Collections.singleton(Action.ADMIN));
+
+    // Bob should now be abe to add/get/delete tag
+    metadataAdmin.addTags(appId, "someTag");
+    metadataAdmin.getTags(MetadataScope.USER, appId);
```
Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java:172]: Verify the tag that you got?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java [hunk 3]
```diff
@@ -72,15 +83,103 @@ public static void setup() throws Exception {
...
+    metadataAdmin.removeTags(appId, "someTag");
```
Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java:173]: Verify that the tag was removed?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java [hunk 4]
```diff
@@ -72,15 +83,103 @@ public static void setup() throws Exception {
...
+
+    // Bob should be able to add/get/delete properties
+    metadataAdmin.addProperties(appId, ImmutableMap.of("key1", "val1"));
+    metadataAdmin.getProperties(MetadataScope.USER, appId);
```
Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java:177]: Verify that the property was added?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java [hunk 5]
```diff
@@ -72,15 +83,103 @@ public static void setup() throws Exception {
...
+    metadataAdmin.removeProperties(appId);
```
Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java:178]: Verify that the property was deleted?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java [hunk 6]
```diff
@@ -72,15 +83,103 @@ public static void setup() throws Exception {
...
+
+    // Bob should be able to get/remove metadata
+    metadataAdmin.getMetadata(MetadataScope.USER, appId);
+    metadataAdmin.removeMetadata(appId);
```
Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataAdminAuthorizationTest.java:182]: Verify that metadata was removed?

**Review by Nishith — COMMENTED** [review]
A few minor test comments. Rest looks good.

**Comment by Nishith** [comment]
After offline discussion, we agreed that we don't need to verify in tests as that is not the concern on authorization.

**Comment by rohitsinha54** [comment]
@Nishith : For the comments about verifying the tags/properties/metadata got added/deleted. I discussed this with terence earlier as my concern was that we are duplicating unit tests since test for verifying logic of above already has unit test. We decided that we should just be expecting UnauthorizedException being thrown and not thrown as from authorization unit test perspective the other stuff is not a concern and is already tested in existing stuff. I have also added a note about this on the class documentation for MetadataAdminAuthorizationTest.java
Lets wait for  @chtyim approval on this and we can decide something accordingly.

**Comment by Nishith** [comment]
LGTM.

**Comment by rohitsinha54** [comment]
will open against develop

**Comment by rohitsinha54** [comment]
Opened against develop : https://github.com/caskdata/cdap/pull/7514

**Pull request closed without merging**